Sources/GRPC/GRPCClientStateMachine.swift

```diff
@@ -0,0 +1,588 @@
+/*
+ * Copyright 2019, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import Foundation
+import NIO
+import NIOHTTP1
+import Logging
+import SwiftProtobuf
+
+enum RecieveResponseHeadError: Error, Equatable {
+  /// The 'content-type' header was missing or the value is not supported by this implementation.
+  case invalidContentType
+
+  /// The HTTP response status from the server was not 200 OK.
+  case invalidHTTPStatus(HTTPResponseStatus)
+
+  /// The encoding used by the server is not supported.
+  case unsupportedMessageEncoding
+
+  /// An invalid state was encountered. This is a serious implementation error.
+  case invalidState
+}
+
+enum SendEndOfRequestStreamError: Error {
+  /// The request stream has already been closed.
+  case endOfStreamCardinalityViolation
+
+  /// An invalid state was encountered. This is a serious implementation error.
+  case invalidState
```

Comment:
I don't think so: `invalidState` really only serves a purpose for testing (since it should be a `preconditionFailure` the rest of the time. The context we add in `InvalidStateError` ~probably~ isn't actually necessary and can be removed.